src/main/java/org/truffleruby/core/ProcessNodes.java

```diff
@@ -10,115 +10,35 @@
-        public RubyNode coerceUnit(RubyNode unit) {
-            return DefaultValueNodeGen.create(floatSecondSymbol, unit);
-        }
-
-        @Specialization(guards = { "isMonotonic(clock_id)", "isRubySymbol(unit)" })
-        protected Object clock_gettime_monotonic(int clock_id, DynamicObject unit) {
-            long time = System.nanoTime();
-            return timeToUnit(time, unit);
-        }
-
-        @Specialization(guards = { "isRealtime(clock_id)", "isRubySymbol(unit)" })
-        protected Object clock_gettime_realtime(int clock_id, DynamicObject unit) {
-            long time = System.currentTimeMillis() * 1_000_000;
-            return timeToUnit(time, unit);
-        }
-
-        @Specialization(guards = { "isThreadCPUTime(clock_id)", "isRubySymbol(unit)" })
-        protected Object clock_gettime_thread_cputime(int clock_id, DynamicObject unit) {
-            return clock_gettime_clock_id(CLOCK_THREAD_CPUTIME_ID, unit);
-        }
-
-        @Specialization(guards = { "isMonotonicRaw(clock_id)", "isRubySymbol(unit)" })
-        protected Object clock_gettime_monotonic_raw(int clock_id, DynamicObject unit) {
-            return clock_gettime_clock_id(CLOCK_MONOTONIC_RAW_ID, unit);
+        @Specialization
+        public long nanoTime() {
+            return System.nanoTime();
         }
 
-        @TruffleBoundary
-        private Object clock_gettime_clock_id(int clock_id, DynamicObject unit) {
-            final ClockGetTime libCClockGetTime = getContext().getNativePlatform().getClockGetTime();
-            TimeSpec timeSpec = new TimeSpec(jnr.ffi.Runtime.getRuntime(libCClockGetTime));
-            int r = libCClockGetTime.clock_gettime(clock_id, timeSpec);
-            if (r != 0) {
-                throw new RaiseException(coreExceptions().systemCallError("clock_gettime failed: " + r, r, this));
-            }
-            long nanos = timeSpec.getTVsec() * 1_000_000_000 + timeSpec.getTVnsec();
-            return timeToUnit(nanos, unit);
-        }
-
-        private Object timeToUnit(long time, DynamicObject unit) {
-            assert RubyGuards.isRubySymbol(unit);
-            if (unit == nanosecondSymbol) {
-                return time;
-            } else if (unit == floatMicrosecondSymbol) {
-                return time / 1e3;
-            } else if (unit == floatSecondSymbol) {
-                return time / 1e9;
-            } else {
-                throw new UnsupportedOperationException(Layouts.SYMBOL.getString(unit));
-            }
-        }
-
-        protected static boolean isMonotonic(int clock_id) {
-            return clock_id == CLOCK_MONOTONIC;
-        }
-
-        protected static boolean isRealtime(int clock_id) {
-            return clock_id == CLOCK_REALTIME;
-        }
+    }
 
-        protected static boolean isThreadCPUTime(int clock_id) {
-            return clock_id == CLOCK_THREAD_CPUTIME;
-        }
+    @Primitive(name = "process_time_currenttimemillis", needsSelf = false)
```

Comment:
I would name those primitive following their purpose, not their implementation, i.e. `process_time_realtime` and `process_time_monotonic`.

Reply:
I'm not sure either of these are monotonic are they? I've named them after what they simply wrap. That way I can't imply anything except for that they call that method.

Reply:
`currentTimeMillis()` is realtime and nanoTime() [is monotonic](https://docs.oracle.com/javase/7/docs/api/java/lang/System.html#nanoTime()) according to its JavaDoc.
We could also just call the native function for both these cases, that would provide better precision for realtime. But probably best to not change that in this PR.

Reply:
I looked at their JavaDoc and can't see what you see. For `nanoTime` are you using 'some fixed but arbitrary origin time' to say that? Seems like if it was monotonic it would actually use that word?

Reply:
Yes, it needs a bit of interpretation :)
`The same origin is used by all invocations of this method in an instance of a Java virtual machine` => monotonic in the process.
But of course the JavaDoc doesn't tell you it uses `clock_gettime`.